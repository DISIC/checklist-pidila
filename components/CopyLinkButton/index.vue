<template>
  <li class="link-tool" v-on-click-outside="hideAlertSuccess">
    <p :role="displayedRoleAttribute" aria-live="assertive">
      <span v-if="displayedAlertSuccess" class="alert-success">
        Le lien a été copié dans le presse-papier.
      </span>
    </p>
    <button
      type="button"
      class="btn btn-tool btn-link"
      title="Copier le lien permanent vers cette liste"
      @click="copyUrl"
      @keyup.enter="copyUrl"
      @blur="hideAlertSuccess"
      @onfocusout="hideAlertSuccess"
      @keydown.esc="hideAlertSuccess">
      <svg class="svg-icon icon-tool icon-link click-through-able" aria-hidden="true" focusable="true">
        <use xlink:href="svg/icon-sprite.svg#link"></use>
      </svg>
      <span class="sr-only">Copier le lien permanent vers cette liste</span>
    </button>
  </li>
</template>

<script>
  export default {
    name: 'CopyLinkButton',
    data: function () {
      return {
        displayedAlertSuccess: false
      }
    },
    methods: {
      displayAlertSuccess: function () {
        this.displayedAlertSuccess = true
      },
      hideAlertSuccess: function () {
        this.displayedAlertSuccess = false
      },
      copyUrl: function (event) {
        var tempInput = document.createElement('input')
        tempInput.style = 'position: absolute; left: -9999px'
        tempInput.value = global.location.href
        tempInput.setAttribute('aria-hidden', 'true')
        tempInput.classList.add('sr-only')
        document.body.appendChild(tempInput)
        tempInput.select()
        document.execCommand('copy')
        document.body.removeChild(tempInput)
        event.target.focus()
        this.displayAlertSuccess()
      }
    },
    computed: {
      displayedRoleAttribute: function () {
        return this.displayedAlertSuccess ? '' : 'presentation'
      }
    }
  }
</script>
