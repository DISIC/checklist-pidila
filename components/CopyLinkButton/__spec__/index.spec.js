import {createLocalVue} from '../../../vue-test.config'
import {shallowMount} from '@vue/test-utils'
import CopyLinkButton from '../index.vue'

describe('components | CopyLinkButton', () => {

  let localVue

  beforeEach(() => {
    document.execCommand = jest.fn()
    localVue = createLocalVue()
  })

  it('should match snapshot', () => {
    // when
    const wrapper = shallowMount(CopyLinkButton)

    // then
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should call the document.execCommand with the argument "copy" on click', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    document.getElementById = () => {
      return {
        select: () => ''
      }
    }

    // when
    wrapper.find('button.btn-link').trigger('click')

    // then
    expect(document.execCommand).toHaveBeenCalledWith('copy')
  })

  it('should call the document.execCommand with the argument "copy" on keymap enter', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    document.getElementById = () => {
      return {
        select: () => ''
      }
    }

    // when
    wrapper.find('button.btn-link').trigger('keyup.enter')

    // then
    expect(document.execCommand).toHaveBeenCalledWith('copy')
  })

  it('should not display a tooltip popover when component is mounted', () => {
    // given
    const $bus = {}

    // when
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    // then
    expect(wrapper.findAll('span.alert-success')).toHaveLength(0)
  })

  it('should display a tooltip popover when "Copier lien" button is clicked', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    // when
    wrapper.find('button.btn-link').trigger('click')

    // then
    expect(wrapper.findAll('span.alert-success')).toHaveLength(1)
  })

  it('should remove the role attribute when copy link button is clicked', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    // when
    wrapper.find('button.btn-link').trigger('click')

    // then
    expect(wrapper.findAll('p[role=presentation]')).toHaveLength(0)
  })

  it('should contain a span named alert-success when copy link button is clicked', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})

    // when
    wrapper.find('button.btn-link').trigger('click')

    // then
    expect(wrapper.findAll('span.alert-success')).toHaveLength(1)
  })

  it('should not contain a span named alert-success when escape is pressed and focus is on button link', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}})
    // when
    wrapper.find('button.btn-link').trigger('click')
    wrapper.find('button.btn-link').trigger('keydown.esc')

    // then
    expect(wrapper.findAll('span.alert-success')).toHaveLength(0)
  })

  it('should not contain a span named alert-success when copy link button lose focus', () => {
    // given
    const $bus = {}
    const wrapper = shallowMount(CopyLinkButton, {localVue, mocks: {$bus}, attachToDocument: true})
    wrapper.find('button.btn-link').trigger('click')

    // when
    wrapper.find('button.btn-link').trigger('blur')

    // then
    expect(wrapper.findAll('span.alert-success')).toHaveLength(0)
  })
})
