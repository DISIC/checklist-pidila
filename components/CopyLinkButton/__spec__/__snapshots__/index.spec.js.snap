// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | CopyLinkButton should match snapshot 1`] = `
<li
  class="link-tool"
>
  <p
    aria-live="assertive"
    role="presentation"
  >
    <!---->
  </p>
   
  <button
    class="btn btn-tool btn-link"
    title="Copier le lien permanent vers cette liste"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-icon icon-tool icon-link click-through-able"
      focusable="true"
    >
      <use
        xlink:href="svg/icon-sprite.svg#link"
      />
    </svg>
     
    <span
      class="sr-only"
    >
      Copier le lien permanent vers cette liste
    </span>
  </button>
</li>
`;
